Reduce types loaded on startup significantly

davidwrighton opened the pull request:
Reduce the number of type loads triggered by complex generic constraints. Experiments show about a 20% reduction in startup time on a proxy of the AWS lambda scenario.

The fixes are:
1. When a method/type has constraints, do not load the constraint types directly if they are generic interfaces. Instead, do accessibility checking and the constraint validation on the signature of the constraint instead.
2. When a type is loaded, enhance the special marker type optimization in the interface map to allow Interfaces which require implementation of other interfaces to use the special marker type. Notably, use the special marker type so that if a generic interface's first type parameter is used as all of the generic type parameters of a required interface we use the special marker type to indicate that the type is loaded.
    - This also forces us to have a somewhat more complicated implementation of casting, since the previous model of special marker types could not experience an equivalent interface match
3. When resolving a default interface method, attempt to avoid loading all the interface types associated with a type. Do so by changing the iteration of interfaces to check to see if the approximate interface could possibly have an implementation
4. When resolving a static virtual method, attempt to avoid loading all the types of the MethodImpl records. Do so by doing a Method name check before loading the type.

In addition, fix a test so that it still throws when loading an invalid type. This required adding a path which forced the interface type to be loaded.

As part of testing this work, the lack of generic variable type safety checks was noticed in crossgen2, and the following changes were made
- Add support for doing circularity checks on type and method generic parameters to the TypeValidationChecker in crossgen2
- Add support for doing variance safety checks to the type loader in crossgen2

And in the runtime
- Remove the circularity of type variables checks happening in generic method load, as it is redundant
- Put all of the variance safety checks in the runtime under the SkipTypeValidation flag as crossgen2 can now do it reliably

Related to the work in #120407

davidwrighton committed:
Rework the  special instantiation type logic to be more effective
- Currently it only handles interfaces defined on valuetypes
- Change it to work for interfaces that are required implementation of other interfaces
  - (In that case, if the interface is generic, the special instantiation type is the first type parameter of the interface, not anything else.)

This change removes some safety checks that I can't find the rationale for. This may cause some entertaining failures in CI

davidwrighton committed:
Reduce constraint loading by not forcing constraints to be loaded eagerly
- Reduce the set of constraints that need to be loaded for Bounds and cast checking

TODO:
There is a path in InitTypeContext which I have #ifdef'd out as I don't understand the comment
The accessibility checking needs full loading now, as we don't have a scheme to just look at all the typedefs in a signature, and instead need to do a load of the full type and work from there. This could be revisited.

davidwrighton committed:
Clear up the issues around GetDeclaringMethodTableFromTypeVarTypeDesc

davidwrighton committed:
Tweak, and add comments...

This is DRAFT as we now have an unused state WhichConstraintsToLoad::TypeOrMethodVarsOnly which we never use. (It could be used for the Bounds algorithm, but that isn't actually used without doing the more expensive logic which does access validation)

davidwrighton committed:
Fix build errors

davidwrighton committed:
Fix bugs making it not work properly

davidwrighton committed:
Fix interface matching bug found in testing

davidwrighton committed:
Optimize virtual static and default interface method resolution logic
- Tweak the logic so that it will attempt to avoid loading types if they are cannot have implementations of the method we're looking for

davidwrighton committed:
Update loader/classloader/generics/ByRefLike/ValidateNegative test
- This test was no longer actually forcing the type which is supposed to throw a TypeLoadException to load within the test
- Tweak the test to actually force the offending type to be loaded

davidwrighton committed:
Pass around the type I want passed around

davidwrighton committed:
Remove WhichConstraintsToLoad::TypeOrMethodVarsOnly

davidwrighton committed:
Fix out of date comment

davidwrighton committed:
Merge branch 'main' into reduce_constraint_load

jkotas reviewed (commented):
The test failure looks related

davidwrighton committed:
- Add support for doing circularity checks on type and method generic parameters to the TypeValidationChecker in crossgen2
- Add support for doing variance safety checks to the type loader in crossgen2
- Remove the circularity of type variables checks happening in generic method load, as it is redundant
- Put all of the variance safety checks in the runtime under the SkipTypeValidation flag as crossgen2 can now do it reliably

davidwrighton committed:
Code review feedback

davidwrighton committed:
- Flesh out implementation of sig parser for constraint accessibility
- Add test for constraints with function pointers in them

jkotas reviewed (commented):
crossgen2 test still failing

davidwrighton committed:
- Add crossgen2 validation of interface impl variance rules

davidwrighton committed:
- Adjust contracts as suggested
- add SkipTypeValidation helper

jkotas reviewed (approved):
LGTM once the tests are passing

mangod9 commented:
/backport to release/10.0-staging
